include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,1049 @@
+
+    /// @copydoc const_iterator
+    using iterator = const_iterator;
+
+    /** A reverse read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+     */
+    using reverse_iterator = std::reverse_iterator<iterator>;
+
+    /// @copydoc reverse_iterator
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    /** Constructor
+
+        Default-constructed objects represent
+        the empty string.
+
+     */
+    BOOST_CONSTEXPR
+    decoded_view(
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(nullptr)
+        , n_(0)
+        , dn_(0)
+        , plus_to_space_(opt.plus_to_space)
+    {}
+
+    /** Constructor
+     */
+    decoded_view(
+        char const* str,
+        pct_decode_opts const& opt = {}) noexcept
+        : decoded_view(string_view(str), opt)
+    {}
+
+    /** Constructor
+     */
+    decoded_view(
+        char const* str,
+        size_type len,
+        pct_decode_opts const& opt = {}) noexcept
+        : decoded_view(string_view(str, len), opt)
+    {}
+
+    /** Constructor
+     */
+#ifndef BOOST_URL_DOCS
+    template <
+        class End,
+        typename std::enable_if<std::is_same<End, char const*>::value, int>::type = 0>
+    BOOST_CXX14_CONSTEXPR
+    decoded_view(
+        char const* first,
+        End last,
+        pct_decode_opts const& opt = {}) noexcept
+#else
+    BOOST_CXX14_CONSTEXPR
+    decoded_view(
+        char const* first,
+        char const* last,
+        pct_decode_opts const& opt = {}) noexcept
+#endif
+        : decoded_view(string_view(first, last), opt)
+    {
+        BOOST_ASSERT(last - first >= 0);
+    }
+
+    /** Constructor
+     */
+    decoded_view(
+        urls::string_view str,
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(str.data())
+        , n_(str.size())
+        , dn_(pct_decode_bytes_unchecked(str))
+        , plus_to_space_(opt.plus_to_space)
+    {
+        error_code ec;
+        auto is_unambiguous = [](unsigned char ch)
+        {
+            return ch != '%';
+        };
+        validate_pct_encoding(str, ec, opt, is_unambiguous);
+        if (ec.failed())
+            detail::throw_invalid_argument(
+                BOOST_CURRENT_LOCATION);
+    }
+
+#if !defined(BOOST_NO_CXX17_HDR_STRING_VIEW)
+    /** Constructor
+     */
+    decoded_view(
+        std::string_view str,
+        pct_decode_opts const& opt = {}) noexcept
+        : decoded_view(urls::string_view(str), opt)
+    {}
+#endif
+
+    /** Returns an iterator to the beginning
+     */
+    constexpr
+    const_iterator
+    begin() const noexcept;
+
+    /// @copydoc begin()
+    constexpr
+    const_iterator
+    cbegin() const noexcept;
+
+    /** Returns an iterator to the end
+     */
+    constexpr
+    const_iterator
+    end() const noexcept;
+
+    /// @copydoc end()
+    constexpr
+    const_iterator
+    cend() const noexcept;
```

Comment:
Why do we need these? no one is calling cbegin and cend on string views... and no one is going to call these on our type. Lets do an experiment, remove the unnecessary members and lets wait to see if anyone notices.